pkg/operation/botanist/secrets.go

```diff
@@ -16,925 +16,852 @@ package botanist
+			if err != nil {
+				errorList = append(errorList, err)
+			}
+		}(s)
+	}
+
+	// Wait and check wether an error occurred during the parallel processing of the Secret creation.
+	wg.Wait()
+	if len(errorList) > 0 {
+		return fmt.Errorf("Errors occurred during secret generation: %+v", errorList)
+	}
+
+	// Create kubeconfig and ssh-keypair secrets also in the project namespace in the Garden cluster
+	for key, value := range map[string]string{"kubeconfig": "kubecfg", "ssh-keypair": "ssh-keypair"} {
+		secretObj := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s.%s", b.Shoot.Info.Name, key),
+				Namespace: b.Shoot.Info.Namespace,
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(b.Shoot.Info, gardenv1beta1.SchemeGroupVersion.WithKind("Shoot")),
+				},
+			},
+			Type: corev1.SecretTypeOpaque,
+			Data: b.Secrets[value].Data,
+		}
+		if _, err := b.K8sGardenClient.CreateSecretObject(secretObj, true); err != nil {
+			return err
+		}
+	}
+
+	// Compute a map of check sum for the data of all our secrets.
+	for name, secret := range b.Secrets {
+		b.CheckSums[name] = computeSecretCheckSum(secret.Data)
+	}
+
+	return nil
+}
+
+// DeployCloudProviderSecret creates or updates the cloud provider secret in the Shoot namespace
+// in the Seed cluster.
+func (b *Botanist) DeployCloudProviderSecret() error {
+	var (
+		checksum = computeSecretCheckSum(b.Shoot.Secret.Data)
+		secret   = &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      common.CloudProviderSecretName,
+				Namespace: b.Shoot.SeedNamespace,
+				Annotations: map[string]string{
+					"checksum/data": checksum,
+				},
+			},
+			Type: corev1.SecretTypeOpaque,
+			Data: b.Shoot.Secret.Data,
+		}
+	)
+
+	if _, err := b.K8sSeedClient.CreateSecretObject(secret, true); err != nil {
+		return err
+	}
+
+	b.Secrets[common.CloudProviderSecretName] = b.Shoot.Secret
+	b.CheckSums[common.CloudProviderSecretName] = checksum
+	return nil
+}
+
+// DeleteGardenSecrets deletes the Shoot-specific secrets from the project namespace in the Garden cluster.
+// TODO: https://github.com/gardener/gardener/pull/353: This can be removed in a future version as we are now using owner
+// references for the Garden secrets (also remove the actual invocation of the function in the deletion flow of a Shoot).
+func (b *Botanist) DeleteGardenSecrets() error {
+	err := b.K8sGardenClient.DeleteSecret(b.Shoot.Info.Namespace, generateGardenSecretName(b.Shoot.Info.Name, "kubeconfig"))
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	err = b.K8sGardenClient.DeleteSecret(b.Shoot.Info.Namespace, generateGardenSecretName(b.Shoot.Info.Name, "ssh-keypair"))
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+func (b *Botanist) deployOpenVPNTLSAuthSecret(existingSecretsMap map[string]*corev1.Secret) error {
+	name := "vpn-seed-tlsauth"
+	if tlsAuthSecret, ok := existingSecretsMap[name]; ok {
```

Comment:
I see this thing reoccur quite often - how about we always check first, whether a secret exists before running the 'worker' code that actually generates it - this way, we could decouple and reuse the code in several places.